File: 김찬우/프로그래머스 /Lv.2/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        
        // 아이디어 
        //  딕션어리로 최종 닉네임을 저장 
        
        HashMap<String,String> map = new HashMap<>();
        int count = 0;
        
        for(String reco : record){
            
            String[] arr = reco.split(" ");
            
            if(arr[0].equals("Change")){
                count++;
            }
            
            if(arr.length == 2){
                continue;
            }
            
            map.put(arr[1],arr[2]);
            
        }
        
        String[] result = new String[record.length-count];
        
        int index = 0;
        
        
        for(int i=0; i<record.length; i++){
            
            String[] arr = record[i].split(" ");
            
            if(arr[0].equals("Enter")){
                result[index] = map.get(arr[1])+"님이 들어왔습니다.";
                index++;
                
            }else if(arr[0].equals("Leave")){
                
                 result[index] = map.get(arr[1])+"님이 나갔습니다.";
                 index++;
                
            }
            
          
        }
        return result;
    }
}

File: 조준장/level2/오픈채팅방.kt
class Solution {
    fun solution(record: Array<String>): Array<String> {
        val user = mutableMapOf<String, String>()
        return record.map{
            val r = it.split(" ")
            when (r[0]) {
                "Enter", "Change" -> user += r[1] to r[2]
            }
            r
        }.asSequence()
            .filter{it[0] != "Change"}
            .map{
                val explanation = when (it[0]) {
                    "Enter" -> "님이 들어왔습니다."
                    "Leave" -> "님이 나갔습니다."
                    else -> {}
                }
                "${user[it[1]]}$explanation"
        }.toList().toTypedArray()
    }
}
